src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+                    projected_task_proxy.get_operators()[abs_op_id].get_ancestor_operator_id(concrete_task.get()));
+            }
+            plan_step.swap(concrete_plan_step);
+        }
+    }
+
+    return utils::make_unique_ptr<Projection>(
+        move(pdb), move(plan), unsolvable);
+}
+
+void CEGAR::print_collection() const {
+    utils::g_log << "[";
+    for (size_t i = 0; i < projection_collection.size(); ++i) {
+        const unique_ptr<Projection> &projection = projection_collection[i];
+        if (projection) {
+            utils::g_log << projection->get_pattern();
+            if (i != projection_collection.size() - 1) {
+                utils::g_log << ", ";
+            }
+        }
+    }
+    utils::g_log << "]" << endl;
+}
+
+void CEGAR::compute_initial_collection() {
+    assert(!goals.empty());
+    for (const FactPair &goal : goals) {
+        add_pattern_for_var(goal.var);
+    }
+
+    if (verbosity >= utils::Verbosity::VERBOSE) {
+        utils::g_log << "initial collection: ";
+        print_collection();
+        utils::g_log << endl;
+    }
+}
+
+bool CEGAR::time_limit_reached(
+    const utils::CountdownTimer &timer) const {
+    if (timer.is_expired()) {
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "time limit reached." << endl;
+        }
+        return true;
+    }
+    return false;
+}
+
+/*
+  TODO: this is a duplicate of State::get_unregistered_successor.
+  The reason is that we may apply operators even though they are not
+  applicable due to ignoring blacklisted violated preconditions.
+*/
+State get_unregistered_successor(
+    const shared_ptr<AbstractTask> &task,
+    const State &state,
+    const OperatorProxy &op) {
+    assert(!op.is_axiom());
+    vector<int> new_values = state.get_unpacked_values();
+
+    for (EffectProxy effect : op.get_effects()) {
+        if (does_fire(effect, state)) {
+            FactPair effect_fact = effect.get_fact().get_pair();
+            new_values[effect_fact.var] = effect_fact.value;
+        }
+    }
+
+    assert(task->get_num_axioms() == 0);
+    return State(*task, move(new_values));
+}
+
+FlawList CEGAR::apply_plan(int collection_index, State &current) const {
+    FlawList flaws;
+    Projection &projection = *projection_collection[collection_index];
+    const vector<vector<OperatorID>> &plan = projection.get_plan();
+    if (verbosity >= utils::Verbosity::VERBOSE) {
+        utils::g_log << "executing plan for pattern "
+                     << projection.get_pattern() << ": ";
+    }
+    for (const vector<OperatorID> &equivalent_ops : plan) {
+        bool step_failed = true;
+        for (OperatorID op_id : equivalent_ops) {
+            OperatorProxy op = task_proxy.get_operators()[op_id];
+
+            /*
+              Check if the operator is applicable. If not, add its violated
+              preconditions to the list of flaws.
+            */
+            bool flaw_detected = false;
```

Comment:
It is again not that simple because `operator_is_applicable` would actually and primarily collect violated preconditions as flaws. It would look something like, using the same approach as suggested above, using a single bool return value and filling flaws passed in as non-const parameter:
```
bool CEGAR::get_violated_preconditions(
    const State &current,
    const OperatorProxy &op,
    int collection_index,
    FlawList &flaws) const {
    bool violated_preconditions = false;
    for (FactProxy precondition : op.get_preconditions()) {
        int var = precondition.get_variable().get_id();

        // Ignore blacklisted variables.
        if (blacklisted_variables.count(var)) {
            continue;
        }

        if (current[precondition.get_variable()] != precondition) {
            violated_preconditions = true;
            flaws.emplace_back(collection_index, var);
        }
    }
    return violated_preconditions;
}

if (!get_violated_preconditions(current, op, collection_index, flaws)) {
    step_failed = false;
    flaws.clear();
    current = get_unregistered_successor(task, current, op);
    break;
}
```
I'm not sure this is better than before, but i'm also not really opposed to it. What do you think? Or do you have better ideas?

Reply:
As the flaws are empty iff violated_preconditions is false, I'd suggest to change this to:

```
FlawList CEGAR::get_violated_preconditions(
    const State &current,
    const OperatorProxy &op,
    int collection_index) {
    FlawList flaws;
    ...
    return flaws;
}

flaws = get_violated_preconditions(current, op, collection_index);
if (!flaws.empty()) {
    ...
}
```

Reply:
Good point. I think I simplified the code further by exiting directly from the loop if a plan step fails. Please let me know if you think this should be amended even further.